### PR TITLE
feat: Free parameter estimation from seed

### PR DESCRIPTION
**Pull request by andiwand**
This is a generalization of the bound version `estimateTrackParamsFromSeed`. From the interface one can see what information is really necessary to estimate parameters
```c++
FreeVector estimateTrackParamsFromSeed(const Vector3& sp0, const Vector3& sp1,
                                       const Vector3& sp2,
                                       const Vector3& bField);

template <typename spacepoint_iterator_t>
FreeVector estimateTrackParamsFromSeed(spacepoint_iterator_t spBegin,
                                       spacepoint_iterator_t spEnd,
                                       const Vector3& bField) {
```
vs
```c++
template <typename spacepoint_iterator_t>
std::optional<BoundVector> estimateTrackParamsFromSeed(
    const GeometryContext& gctx, spacepoint_iterator_t spBegin,
    spacepoint_iterator_t spEnd, const Surface& surface, const Vector3& bField,
    ActsScalar bFieldMin, const Acts::Logger& logger = getDummyLogger()) {
```

Originally I wanted to make use of this new version inside the bound version but I ended up doing a different error handling which is also more generic IMO.

We can unify this with the next breaking change. I would propose to drop the bound version completely and let the user deal with the global to local conversion which makes it very explicit that one needs a surface to put the parameters on. A propagator can then be used to extrapolate parameters to a different reference surface.

### CODE REVIEW

**Comment by andiwand** [comment]
cc @timadye @paulgessinger @noemina @pbutti @CarloVarni

File: Core/include/Acts/Seeding/EstimateTrackParamsFromSeed.hpp [hunk 1]
```diff
@@ -20,9 +20,90 @@
...
+///
+/// @param spBegin is the begin iterator for the space points
+/// @param spEnd is the end iterator for the space points
+/// @param bField is the magnetic field vector
+///
+/// @return optional bound parameters
+template <typename spacepoint_iterator_t>
+FreeVector estimateTrackParamsFromSeed(spacepoint_iterator_t spBegin,
+                                       spacepoint_iterator_t spEnd,
+                                       const Vector3& bField) {
```
Comment (CarloVarni) [inline Core/include/Acts/Seeding/EstimateTrackParamsFromSeed.hpp:75]: what about
```suggestion
template <std::ranges::range external_collection_t>
FreeVector estimateTrackParamsFromSeed(external_collection_t collection,
                                       const Vector3& bField) {
```

File: Core/include/Acts/Seeding/EstimateTrackParamsFromSeed.hpp [hunk 2]
```diff
@@ -20,9 +20,90 @@
...
+///
+/// @param sp0 is the bottom space point
+/// @param sp1 is the middle space point
+/// @param sp2 is the top space point
+/// @param bField is the magnetic field vector
+///
+/// @return optional bound parameters
+FreeVector estimateTrackParamsFromSeed(const Vector3& sp0, const Vector3& sp1,
+                                       const Vector3& sp2,
+                                       const Vector3& bField);
```
Comment (CarloVarni) [inline Core/include/Acts/Seeding/EstimateTrackParamsFromSeed.hpp:50]: is anybody using this?

File: Core/include/Acts/Seeding/EstimateTrackParamsFromSeed.hpp [hunk 3]
```diff
@@ -20,9 +21,87 @@
...
+/// @param spBegin is the begin iterator for the space points
+/// @param spEnd is the end iterator for the space points
+/// @param bField is the magnetic field vector
+///
+/// @return the free parameters
+template <std::ranges::range spacepoint_range_t>
+FreeVector estimateTrackParamsFromSeed(spacepoint_range_t spRange,
+                                       const Vector3& bField) {
+  // Check the number of provided space points
+  std::size_t numSP = std::distance(spRange.first, spRange.second);
```
Comment (CarloVarni) [inline Core/include/Acts/Seeding/EstimateTrackParamsFromSeed.hpp:79]:
```suggestion
  std::size_t numSP = spRange.size();
```

File: Core/include/Acts/Seeding/EstimateTrackParamsFromSeed.hpp [hunk 4]
```diff
@@ -20,9 +21,87 @@
...
+      std::nullopt, std::nullopt, std::nullopt};
+  // The first, second and third space point are assumed to be bottom, middle
+  // and top space point, respectively
+  for (auto [isp, sp] : enumerate(spRange)) {
+    if (sp == nullptr) {
+      throw std::invalid_argument("Empty space point found.");
+    }
+    spGlobalPositions[isp] = Vector3(sp->x(), sp->y(), sp->z());
+    spGlobalTimes[isp] = sp->t();
+  }
```
Comment (CarloVarni) [inline Core/include/Acts/Seeding/EstimateTrackParamsFromSeed.hpp:99]:
```suggestion
  std::size_t counter = 0ul;
  for (const auto* sp : spRange) {
    if (sp == nullptr) {
      throw std::invalid_argument("Empty space point found.");
    }
    spGlobalPositions[counter] = Vector3(sp->x(), sp->y(), sp->z());
    spGlobalTimes[counter] = sp->t();
    ++counter;
  }
```

File: Core/include/Acts/Seeding/EstimateTrackParamsFromSeed.hpp [hunk 5]
```diff
@@ -20,9 +21,87 @@
...
+///
+/// @tparam spacepoint_iterator_t  The type of space point iterator
+///
+/// @param spBegin is the begin iterator for the space points
+/// @param spEnd is the end iterator for the space points
+/// @param bField is the magnetic field vector
+///
+/// @return the free parameters
+template <std::ranges::range spacepoint_range_t>
+FreeVector estimateTrackParamsFromSeed(spacepoint_range_t spRange,
```
Comment (CarloVarni) [inline Core/include/Acts/Seeding/EstimateTrackParamsFromSeed.hpp:76]: what about adding also some requirements?
```suggestion
template <std::ranges::range spacepoint_range_t>
requires requires(typename spacepoint_range_t::value_type sp) {
 { sp.x() } -> std::convertible_to<Acts::ActsScalar>;
  { sp.y() } -> std::convertible_to<Acts::ActsScalar>;
   { sp.z() } -> std::convertible_to<Acts::ActsScalar>;
   { sp.t() } -> std::same_as<std::optional<double>>;
}
FreeVector estimateTrackParamsFromSeed(spacepoint_range_t spRange,
```

Comment (CarloVarni) [reply Core/include/Acts/Seeding/EstimateTrackParamsFromSeed.hpp:76]: not sure about this right now ... value_type may be a pointer as well ...

**Comment by CarloVarni** [comment]
@andiwand can you check apogee and Athena is ok with it? I'm afraid of the require statement here.
In case, we can drop it (I know I'm the one who suggested it ... )

**Review by CarloVarni — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [4c9dbcd4079037ac248da51ed8fb657513e68afd]
### :red_circle: Some tests have failed!
Please investigate the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/8539717)!

|status|job|report|
|------|---|------|
|:red_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46159634)||
|:green_circle: | [report_pull_request](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152221)||
|:green_circle: | [run_art_test: test_run4_acts_vertex_PU200](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152220) | [:paperclip:](https://acts-athena-outputs.web.cern.ch/test_run4_acts_vertex_PU200/pulls/3832/)|
|:red_circle: | [run_art_test: test_run4_ttbar_PU200](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152219)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152218) | [:paperclip:](https://acts-athena-outputs.web.cern.ch/test_data18_13TeV_1000evt/pulls/3832/)|
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152216)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152215)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152214)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152213)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152212)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152211)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflowHgtd](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152210)||
|:green_circle: | [test_ActsCheckObjectCountsCachedWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152209)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152208)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152207)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152206)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152205)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152204)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152203)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152202)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152201)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152200)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152199)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152198)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152197)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152196)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152195)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152194)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152193)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152192)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152191)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152190)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152189)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152188)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152187)||
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46152186)||

**Pull request merged**